src/codec/encode.jl

```diff
@@ -4,67 +4,55 @@ function encode_tag(io::IO, field_number, wire_type::WireType)
         initpos = position(io)
-        truncate(io, initpos + MAX_LENGTH_VARINT_BYTES) # 1.
-        seek(io, initpos + MAX_LENGTH_VARINT_BYTES)
-        f(sink, x, V) # e.g. _encode(io, x, Val{:zigzag}) # 2.
+        # We need to encode the encoded size of x before we know it. We first preallocate 1
+        # byte as that is the mininum size of the encoded size.
+        # If our guess is right, it will save us a copy, but we never want to preallocate too much
+        # space for the size, because then we risk outgrowing the buffer that was allocated with exact size
+        # needed to contain the message.
+        encoded_size_len_guess = 1
+        truncate(io, initpos + encoded_size_len_guess)
+        seek(io, initpos + encoded_size_len_guess)
+        # Now we can encode the object itself
+        f(sink, x, V...) # e.g. _encode(io, x) or _encode(io, x, Val{:zigzag})
         endpos = position(io)
-        data_len = endpos - initpos - MAX_LENGTH_VARINT_BYTES
-        seek(io, initpos)                  # 3.
-        vbyte_encode(io, UInt32(data_len)) # --||--
-        lenght_len = position(io) - initpos
-        unsafe_copyto!(io.data, initpos + lenght_len + 1, io.data, initpos + MAX_LENGTH_VARINT_BYTES + 1, data_len) # 4.
-        seek(io, initpos + lenght_len + data_len)
-        truncate(io, initpos + lenght_len + data_len)
+        encoded_size = endpos - initpos - encoded_size_len_guess
+        encoded_size_len = _encoded_size(UInt32(encoded_size))
+        @assert (initpos + encoded_size_len + encoded_size) <= io.maxsize
+        # If our initial guess on encoded size of the size was wrong, then we need to move the encoded data
+        if encoded_size_len_guess < encoded_size_len
+            truncate(io, initpos + encoded_size_len + encoded_size)
+            # Move the data right after the correct size
+            unsafe_copyto!(
+                io.data,
+                initpos + encoded_size_len + 1,
+                io.data,
+                initpos + encoded_size_len_guess + 1,
+                encoded_size
+            )
+        end
+        # Now we can encode the size
+        seek(io, initpos)
+        vbyte_encode(io, UInt32(encoded_size))
+        seek(io, initpos + encoded_size_len + encoded_size)
     else
-        vbyte_encode(io, UInt32(_encoded_size(x, V)))
-        f(sink, x, V)
+        # TODO: avoid quadratic behavior when estimating encoded size by providing a scratch buffer
+        vbyte_encode(io, UInt32(_encoded_size(x, V...)))
+        f(sink, x, V...)
     end
     return nothing
 end
 
-@inline function _with_size(f, io::IO, sink, x)
-    vbyte_encode(io, UInt32(_encoded_size(x)))
-    f(sink, x)
-    return nothing
-end
-
-@inline function _with_size(f, io::IO, sink, x, V)
-    vbyte_encode(io, UInt32(_encoded_size(x, V)))
-    f(sink, x, V)
+@inline function _with_size(f, io::IO, sink, x, V...)
+    # TODO: avoid quadratic behavior when estimating encoded size by providing a scratch buffer
```

Comment:
Not related to this PR, but what is the quadratic behaviour here, and how would a scratch buffer help?

Reply:
We're calling `_encoded_size` recursively for each message field that encodes its length, so deeply nested messages would call this method many times redundantly for the leaf fields etc. By having a scratch buffer, we could do exactly the same thing as we do for seekable IOBuffers -- encoding the data first and then noting their length. We'd end up copying the data from the scratch buffer to the IO object though and carrying the memory overhead of the scratch buffer, so it's not going to be a silver bullet.

Reply:
I see, thanks for the context 👍